### ISSUE TITLE
Wrong behavior with Petsc 3.9.* in //

### ISSUE THREAD

**Issue by vincentchabannes** [feelpp/feelpp#1166]
I noticed that use petsc 3.9.0 or 3.9.1 give troubles in some applications in // (seems good in sequential). I see that only with nonlinear solver.
For example, with thermoelectric toolbox : 
* in linear, the result is OK : 
```mpirun -np 4 ./toolboxes/thermoelectric/feelpp_toolbox_thermoelectric_3d --config-file ../feelpp/toolboxes/thermoelectric/ElectroMagnets/HL-31_H1/HL-31_H1.cfg```
* in non linear, the results is BAD : 
```mpirun -np 4 ./toolboxes/thermoelectric/feelpp_toolbox_thermoelectric_3d --config-file ../feelpp/toolboxes/thermoelectric/ElectroMagnets/HL-31_H1/HL-31_H1_nonlinear.cfg```

We can see with the paraview visualization that something is wrong at the interprocess values.

Tested on MacOS and atlas cluster(ubuntu 16.04).

No problem with petsc 3.8.3

**Comment by vincentchabannes** [comment]
Results of testsuite on atlas with petsc 3.9.1 et slepc 3.9.1 : 
```
Label Time Summary:
testalg              =  71.78 sec (22 tests)
testcore             = 120.82 sec (42 tests)
testcrb              = 194.37 sec (15 tests)
testdiscr            = 653.61 sec (111 tests)
testfilters          =  82.58 sec (14 tests)
testfit              =  13.33 sec (4 tests)
testintegration      =  52.46 sec (14 tests)
testinterpolation    = 126.52 sec (25 tests)
testleaks            =  53.11 sec (16 tests)
testls               =  10.17 sec (3 tests)
testmaterial         =   5.26 sec (2 tests)
testmath             =   5.26 sec (2 tests)
testmesh             =  96.78 sec (26 tests)
testmodels           =   6.70 sec (2 tests)
testopt              =  11.89 sec (4 tests)
testpde              =  19.62 sec (4 tests)
testpoly             =  70.19 sec (18 tests)
testts               =  21.86 sec (7 tests)
testvf               = 136.28 sec (38 tests)

Total Test time (real) = 1759.78 sec

The following tests did not run:
        185 - feelpp_test_space_0d-np-6 (Skipped)
        186 - feelpp_test_space_0d-np-1 (Skipped)
        298 - feelpp_test_bdf3-np-1 (Skipped)
        307 - feelpp_test_laplacianv-np-6 (Skipped)
        308 - feelpp_test_laplacianv-np-1 (Skipped)
        309 - feelpp_test_laplacian-np-6 (Skipped)
        310 - feelpp_test_laplacian-np-1 (Skipped)
        311 - feelpp_test_laplaciant-np-6 (Skipped)
        312 - feelpp_test_laplaciant-np-1 (Skipped)
        323 - feelpp_test_convolve-np-6 (Skipped)

The following tests FAILED:
        257 - feelpp_test_eigs-np-6 (Failed)
        258 - feelpp_test_eigs-np-1 (Failed)
```

The error with feelpp_test_eigs is  : 
```
258: [0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
258: [0]PETSC ERROR: Shift-and-invert requires a target 'which' (see EPSSetWhichEigenpairs), for instance -st_type sinvert -eps_target 0 -eps_target_magnitude
258: [0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
258: [0]PETSC ERROR: Petsc Release Version 3.9.1, Apr, 29, 2018 
258: [0]PETSC ERROR: /ssd/chabannes/Feelpp/work_remotedata/feelpp.build/testsuite/feeldiscr/feelpp_test_eigs on a  named atlas2 by chabannes Fri May 18 10:19:16 2018
258: [0]PETSC ERROR: Configure options --with-shared-libraries=1 --prefix=/data/software/install/petsc/3.9.1/gcc-6.4.0/openmpi-1.10.7 --with-cc=/data/software/install/openmpi/1.10.7/gcc-6.4.0/bin/mpicc --with-cxx=/data/software/install/openmpi/1.10.7/gcc-6.4.0/bin/mpic++ --with-fc=/data/software/install/openmpi/1.10.7/gcc-6.4.0/bin/mpif90 --with-mpiexec=/data/software/install/openmpi/1.10.7/gcc-6.4.0/bin/mpiexec --download-suitesparse=1 --download-ml --download-blacs --download-scalapack --download-fblaslapack --download-mumps --download-hypre --download-ptscotch --with-hdf5=1 --with-hdf5-dir=/data/software/install/hdf5/1.10.1/gcc-6.4.0/openmpi-1.10.7 --with-debugging=0 --COPTFLAGS=-O3 --CXXOPTFLAGS=-O3 --FOPTFLAGS=-O3
258: [0]PETSC ERROR: #1 EPSSetUp() line 272 in /data/software/sources/slepc/3.9.1/gcc-6.4.0/openmpi-1.10.7/slepc-3.9.1-build/slepc-3.9.1/src/eps/interface/epssetup.c
258: [0]PETSC ERROR: #2 EPSSolve() line 135 in /data/software/sources/slepc/3.9.1/gcc-6.4.0/openmpi-1.10.7/slepc-3.9.1-build/slepc-3.9.1/src/eps/interface/epssolve.c
258: [0]PETSC ERROR: #3 solve() line 398 in /ssd/chabannes/Feelpp/work_remotedata/feelpp/feel/feelalg/solvereigenslepc.cpp
```

Thus, the testsuite does show my issue with petsc, but another one with slepc 3.9.*, if someone has an idea to solve  ?

**Comment by romainhild** [comment]
@vincentchabannes I guess that for the slepc error, when usign a target, the position of the spectrum need to be `EPS_TARGET_MAGNITUDE`.
So maybe we should add `ierr = EPSSetWhichEigenpairs ( M_eps, EPS_TARGET_MAGNITUDE );`
to this function:
https://github.com/feelpp/feelpp/blob/6a5cfce937ccdf0bd24112c11269946e3ab9ad5b/feel/feelalg/solvereigenslepc.cpp#L846-L857

**Comment by vincentchabannes** [comment]
the bug with the thermoelectric case has been fixed in faa55a0028d4a529f7cce05434d9c4eaa0067c8f.
